diesel/src/query_builder/mod.rs

```diff
@@ -73,15 +73,37 @@ impl<'a, T: Query> Query for &'a T {
     type SqlType = T::SqlType;
 }
 
+#[doc(hidden)]
+#[allow(missing_debug_implementations)]
+pub enum AstPass<'a, DB> where
+    DB: Backend,
+    DB::BindCollector: 'a,
+{
+    CollectBinds(&'a mut DB::BindCollector),
+    IsSafeToCachePrepared(&'a mut bool),
+}
+
 /// An untyped fragment of SQL. This may be a complete SQL command (such as
 /// an update statement without a `RETURNING` clause), or a subsection (such as
 /// our internal types used to represent a `WHERE` clause). All methods on
 /// [`Connection`](../connection/trait.Connection.html) that execute a query require this
 /// trait to be implemented.
 pub trait QueryFragment<DB: Backend> {
     fn to_sql(&self, out: &mut DB::QueryBuilder) -> BuildQueryResult;
-    fn collect_binds(&self, out: &mut DB::BindCollector) -> QueryResult<()>;
-    fn is_safe_to_cache_prepared(&self) -> bool;
+    fn walk_ast(&self, pass: &mut AstPass<DB>) -> QueryResult<()>;
+
+    fn collect_binds(&self, out: &mut DB::BindCollector) -> QueryResult<()> {
+        self.walk_ast(&mut AstPass::CollectBinds(out))
+    }
+
+    fn is_safe_to_cache_prepared(&self) -> bool {
+        let mut result = true;
+        match self.walk_ast(&mut AstPass::IsSafeToCachePrepared(&mut result)) {
+            Ok(()) => result,
+            // This should never fail, but let's avoid panicking if we somehow do
+            Err(_) => false,
```

Comment:
What do you suggest? The only other options I could come up with are printing to stderr or panicking. The first is untenable, and I've become more hesitant towards the second. Ultimately in this particular case the worst that could happen is we skip prepared statement caching, which is why I went this path. Interested to hear your thoughts.